compiler/rustc_expand/src/proc_macro_server.rs

```diff
@@ -421,13 +425,58 @@ impl server::TokenStream for Rustc<'_> {
         parse_stream_from_source_str(
             FileName::proc_macro_source_code(src),
             src.to_string(),
-            self.sess,
+            self.sess(),
             Some(self.call_site),
         )
     }
     fn to_string(&mut self, stream: &Self::TokenStream) -> String {
         pprust::tts_to_string(stream)
     }
+    fn expand_literal(&mut self, stream: &Self::TokenStream) -> Result<Self::Literal, ()> {
+        // Parse the expression from our tokenstream.
+        let expr = {
+            let mut p = rustc_parse::stream_to_parser(
+                self.sess(),
+                stream.clone(),
+                Some("proc_macro expand literal"),
+            );
+            let expr = p.parse_expr().map_err(|mut err| {
+                err.cancel();
+            })?;
+            if p.token != token::Eof {
+                return Err(());
+            }
+            expr
+        };
+
+        // Perform eager expansion on the expression.
+        let expr = self
+            .ecx
+            .expander()
+            .fully_expand_fragment(crate::expand::AstFragment::Expr(expr))
+            .make_expr();
+
+        // Extract the literal value from the expression.
+        let lit = match &expr.kind {
+            ast::ExprKind::Lit(l) => l.token,
+            ast::ExprKind::Unary(ast::UnOp::Neg, e) => match &e.kind {
+                ast::ExprKind::Lit(l) => match l.token {
+                    t @ token::Lit { kind: token::Integer | token::Float, .. } => {
+                        token::Lit::new(t.kind, Symbol::intern(&format!("-{}", t.symbol)), t.suffix)
+                    }
+                    _ => return Err(()),
+                },
+                _ => return Err(()),
+            },
+            _ => return Err(()),
+        };
+
+        // Filter to valid literal kinds to return from this method.
+        match lit.kind {
+            token::Bool | token::Err => Err(()),
+            _ => Ok(Literal { lit, span: expr.span }),
```

Comment:
Using the literal span (`ast::ExprKind::Lit(l) => l.span`) instead of the expr span would be more correct here (they can potentially differ).